### PR TITLE
Engine: implemented command line options for config and user dir paths

### PR DESCRIPTION
**Pull request by ivan-mogilko**
In response to #961 and #975.

Implemented following:

 --conf - defines explicit and only config file path;
 --user-data-dir - defines user data directory for game files (aka $SAVEGAMEDIR$);
 --shared-data-dir - defines shared data directory for game files (aka $APPDATADIR$);

Example of use:

`ags --conf ./my.cfg --user-data-dir ./saves --shared-data-dir ./appdata`

NOTES:

`--conf` forces engine to ignore all other config files.

`--user-data-dir` and `--shared-data-dir` override corresponding options from config file.

---

The biggest issue I have with `--conf` is that so far it does not have a clear rule in regards to *saving config back*. In other words, must it be considered a read-only special config, or actually "user config" with all the rules applying?
A while ago I've disabled saving config by the engine on exit because it was confusing for number of reasons (see #1156, also related to #975).
But, if we ever enable that, or, for instance, let game developers change and save config from within the game script, what should be the rule in regards to file passed as `--conf`?
I have a concern that if unresolved now then in the long run this may become a cause of more logical conflicts.

### CODE REVIEW

**Comment by rofl0r** [comment]
> The biggest issue I have with --conf is that so far it does not have a clear rule in regards to saving config back. 

i suspect if the user uses command line option to force a specific option, it would be unexpected to have it overwritten by engine.
for example i was surprised when AGS started last game i played in fullscreen, even though i put windowed manually into config, probably because it was in fullscreen when i quitted it. i'd rather that windowed stays the default unless i change it.

**Comment by ivan-mogilko** [comment]
Well, what I am trying to say, this option in pr, it tells to explicitly use certain config for startup. That's fine on its own, but this does not cover another use case when players wanted not only read but also write config in game dir. Because I disabled writing config by engine (on exit), right now this may happen only with embedded setup program (on Windows), or any standalone setup app that would use engine's rules of finding user config.
Assuming in the future we may add support for writing config by script command, that would also cause similar problem.

On a somewhat related note, right now there's already at least one script command that may change game setup: it's the System.Windowed that lets game switch from windowed to fullscreen. Making abstract assumption that a game uses a custom config file, in theory it may even switch on start. I guess that's a separate issue, but wanted to mention this, because there is also an old feature request to support other graphic settings in script.

**Comment by ivan-mogilko** [comment]
@legluondunet , could you tell, given the above does current solution work for you?

I've been also thinking maybe add an option that tells to use default config's location (in game dir) as user config location in all operations (so long as it's writeable).

**Comment by legluondunet** [comment]
1) About read/write config, I think it will be better if ags would write options in config file on the fly.
2) Another thing, what's the matter if I start  a game ike this:
ags -conf ./game.cfg
and game.cfg file does not exit, ags will create it with default ags values right?
3) What's the matter if the folder where is located config/saves files is read only?

**Comment by ivan-mogilko** [comment]
> 1. About read/write config, I think it will be better if ags would write options in config file on the fly.

Could you please elaborate, what situation you are refering to?

> 2. Another thing, what's the matter if I start  a game ike this:
>    ags -conf ./game.cfg
>    and game.cfg file does not exit, ags will create it with default ags values right?

The latest engine does not create or write anything.
If we take slightly earlier version, it wrote only *certain* settings on exit (the settings that could be changed by script commands).
If we take embedded setup program (it works only on Windows), that would create user config with default values, but that config would be located according to standard path rules.
In its current implementation `--conf` does not directly correspond to user config, it does not replace its location; that's why I proposed to add another option that actually would do that, either instead or along with `--conf`.

> 3. What's the matter if the folder where is located config/saves files is read only?

Back when the engine was writing config on exit it would silently fail. The embedded setup would display an error message.

Any changes to above behavior may be a subject of discussion of course.

**Comment by legluondunet** [comment]
In summary if the config file is read in read only that suits me very well, as long as some games does not complain.
If read-only config file cause an issue for some games, it will be another issue to open.

**Comment by ivan-mogilko** [comment]
Well, I'm still not fully certain if we have same understanding of this.
Right now `--conf` does not impose any additional rules. Like, it does not tell engine "not to write into this file under any circumstances". It only tells "read config from there on startup, and from nowhere else".
If we assume that engine will be writing user config again at some point, and if you pass a file that is NOT the one which engine or setup program is considering user config location, then they won't write in it.


On another note, I have a dumb question. Does it matter if the command options for user data and shared data dir have precisely same name as in config file? I just realized in config they have underscores between name parts, but I made cmd args with dashes. Actually, we have a bit of a mess there, not all command arg names correspond to config 1 to 1.

**Comment by rofl0r** [comment]
> Does it matter if the command options for user data and shared data dir have precisely same name as in config file? I just realized in config they have underscores between name parts, but I made cmd args with dashes.

traditional command line tools all use dashes instead of underline for wordbreaks, so that sounds good as is.

**Comment by ivan-mogilko** [comment]
@legluondunet I am sorry, I should have checked your older comments in #975 to recall your use case.

Basically you have a game platform inside Linux, and are using your own setup utility to write game config inside game directory.
Your problem was that engine was writing a user config in user .local/share etc, overriding your settings.
This should not longer happen as of [AGS 3.5.0.31](https://github.com/adventuregamestudio/ags/releases/tag/v.3.5.0.31), where I disabled engine writing config on exit. Idea is that it is disabled *temporarily* until there's a good method to control this behavior, such as a restricting where it saves this config.

I made a mistake having several tickets refering to related problems but not updating them in time.

So, what I will do now, I will change this PR to only add `--user-data-dir` and `--shared-data-dir` as these correspond to existing options and may be still useful. For now I will not be adding any `--conf` or similar options, because these are simply not necessary (at least not on Linux where we do not have an embedded setup program).

The problem with writing config back remains (it still exists on Windows), but I will address this separately.

**Comment by legluondunet** [comment]
I must admit that I have a hard time following you, in the end you are not going to add a command line option to tell AGS where the config file is located? 
Like
ags --config-path /your/path/to/config.file
or
ags --config- /your/path/to/config.file
or 
ags --base-path  /your/path/to/config.file and savefiles
?

**Comment by ivan-mogilko** [comment]
> in the end you are not going to add a command line option to tell AGS where the config file is located?

The question is, what is the use case? Do you actually need this ability to tell engine explicit config location? Is that config is outside of game directory somewhere for example?

**Comment by ivan-mogilko** [comment]
> I must admit that I have a hard time following you, in the end you are not going to add a command line option to tell AGS where the config file is located?

Sorry, let me rephrase this. Yes, I am not going to add this command line option, because I came to conclusion that it's not necessary in your case, as your problem was related to engine writing user config in a different location. But because that behavior is disabled this should no longer be a problem.

I may be mistaken (again) of course, in which case I ask to elaborate on the situation.

**Comment by legluondunet** [comment]
I need to understand...
So after this PR applied, where AGS will search for game config file? 
Where it will search for savefiles?
English is not my native language, thank you for your understanding.

**Comment by ivan-mogilko** [comment]
It will be the same. My question was, do I need to change that?

Because the latest engine _does not_ write any of these configs. So the only way they may be created is a) by user's hand b) by some setup application.

**Comment by ivan-mogilko** [comment]
> Where it will search for savefiles?

By default it searches for them in the user's folder. But you may also set "user_data_dir" in config. After this PR you may also pass `--user-data-dir` in command line to define the save game location.

**Comment by ivan-mogilko** [comment]
Alright. I made a mistake.

Perhaps you have a need to restrict engine to very particular config. So I will just add such command. Then if this would conflict with new engine behavior or some games later we may find a solution.

I apologize for all the confusion; I am thinking about several problems at a time.

**Comment by legluondunet** [comment]
>Because the latest engine does not write any of these configs anymore. So the only way they may be created is a) by user's hand b) by some setup application.

ok, let's go for this scenari, should be enough to avoid config file confusion for users.
thank you to take time to explain me in details.

**Comment by ivan-mogilko** [comment]
Ok.

I also apologize for all the confusion, I was thinking about several problems at a time.

**Comment by rofl0r** [comment]
> For now I will NOT be adding any --conf or similar options, because these are simply not necessary (at least not on Linux where we do not have an embedded setup program).

what? that logic seems to be reversed, especially because we dont have a setup program we may want to pass the config file location to engine (for example using a *custom* frontend)

i'm still interested into having the --config option so i can finally write a wrapper around ags that always uses the config from current directory and i can finally stop to go hunting for the right config file.

**Comment by ivan-mogilko** [comment]
My point above was that unless you create a config in user location, the engine will use the one in the game dir anyway. There would not be a need to hunt for config elsewhere.

But yes, I can add the command for explicit config location. The question that I've been asking is: should it be just a config for reading at startup, or it should replace user config in all aspects. As in, assuming engine would like to write config back, should it use the path provided with this command?

**Comment by ivan-mogilko** [comment]
Added `--conf` f47c365473abe9ffd4e269c563e9bb14ff428985

**Comment by rofl0r** [comment]
> As in, assuming, hypothetically, engine would like to write a modified config back, should it use the path provided with this command?

i still fail to see the purpose of having the engine write a config back, if not to document what its default values are.

but in case this gets the default again we could add a switch --config-readonly (boolean switch) or --ro-config-path=./acsetup.cfg which sets both readonly to true and sets the path.

**Comment by ivan-mogilko** [comment]
> i still fail to see the purpose of having the engine write a config back

For instance, there was a feature request of editing config from within the game, so that if player changes settings in game menus it would be saved and restored on next run.

I should also probably mention that since 3.5.0 there's now `--tell-configpath` command that prints locations of all configs as expected by the engine to stdout. In theory this may be used by setup program if it needs to follow engine's default rules.

**Comment by rofl0r** [comment]
`--tell-configpath` is an improvement and i was actually refering to that when i talked about `hunting down` the location, because you always have to look into all 3 configs to figure out from where the option you don't want comes from...

> so that if player changes settings in game menus it would be saved and restored on next run.

hmm that sounds as if the option they actually wanted is a script function called SaveConfigToDisk()
